presto-main/src/main/java/com/facebook/presto/execution/SharedBuffer.java

```diff
@@ -212,12 +208,19 @@ public synchronized void setOutputBuffers(OutputBuffers newOutputBuffers)
             TaskId bufferId = entry.getKey();
             if (!namedBuffers.containsKey(bufferId)) {
                 checkState(state.get().canAddBuffers(), "Cannot add buffers to %s", SharedBuffer.class.getSimpleName());
-                NamedBuffer namedBuffer = new NamedBuffer(bufferId, entry.getValue());
+                PagePartitionFunction partitionFunction = entry.getValue();
+                int partition = 0;
+                if (partitionFunction instanceof HashPagePartitionFunction) {
+                    partition = ((HashPagePartitionFunction) partitionFunction).getPartition();
+                }
+                NamedBuffer namedBuffer = new NamedBuffer(bufferId, partition);
                 // the buffer may have been aborted before the creation message was received
                 if (abortedBuffers.contains(bufferId)) {
                     namedBuffer.abort();
                 }
                 namedBuffers.put(bufferId, namedBuffer);
+                partitionToPageBuffer.computeIfAbsent(partition, k -> new PageBuffer(maxBufferedBytes));
+                partitionToNamedBuffer.computeIfAbsent(partition, k -> new HashSet<>()).add(namedBuffer);
```

Comment:
It is hard to see the `add` ad the end of this line. Assign the Set to a variable.